circuit-lib/circuitlib-rs/src/gnark/helpers.rs

```diff
@@ -11,46 +11,65 @@ use serde_json::json;
 
 use crate::gnark::constants::{HEALTH_CHECK, SERVER_ADDRESS};
 
-pub fn spawn_gnark_server(path: &str, wait_time: u64) -> Child {
-    let server_process = Command::new("sh")
-        .arg("-c")
-        .arg(path)
-        .spawn()
-        .expect("Failed to start server process");
-
-    // Wait for the server to launch before proceeding.
-    thread::sleep(Duration::from_secs(wait_time));
+static IS_LOADING: AtomicBool = AtomicBool::new(false);
 
-    server_process
+pub async fn spawn_gnark_server(
+    path: &str,
+    restart: bool,
+    inclusion: bool,
+    non_inclusion: bool,
+    combined: bool,
+) {
+    if restart {
+        kill_gnark_server();
+    }
+    if !health_check(1, 3).await && !IS_LOADING.load(Ordering::Relaxed) {
+        IS_LOADING.store(true, Ordering::Relaxed);
+        let mut arg: String = "".to_string();
+        if inclusion {
+            arg += "inclusion";
+        } else if non_inclusion {
+            arg += "non-inclusion";
+        } else if combined {
+            arg += "combined";
+        }
```

Comment:
If we decide to keep them mutually exclusive and introduce an enum, we could change it to

```suggestion
        let arg = match proof_type {
            ProofType::Inclusion => "inclusion",
            ProofType::NonInclusion => "non-inclusion",
            ProofType::Combined => "combined",
        };
```

If we don't, then we probably don't want to do `else if`, but rather include all enabled options like

```suggestion
        if inclusion {
            arg += "inclusion ";
        }
        if non_inclusion {
            arg += "non-inclusion ";
        }
        if combined {
            arg += "combined";
        }
```